File: problem3.java
"""
    Explanation: I used binary seach to find the target. The binary search finds the mid point first,
    then the mid point is compared with the target to get the value. Edge cases added for checking
    whether the first / last element is the target.
    Time Complexcity: O (log n)
    Space Complexcity: O(1)
""";


class Solution {
    public int search(ArrayReader reader, int target) {

        int start = 0, end = Integer.MAX_VALUE;

        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (target == reader.get(mid)) {
                return mid;
            } else if (target < reader.get(mid)) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        if (reader.get(start) == target) {
            return start;
        } else if (reader.get(end) == target) {
            return end;
        }
        return -1;
    }
}

File: Problem-3.js
//Ran Successfully on leetCode - solution accepted
//Time: O(logk), where k is an index of target value, Space O(1)
var search = function (reader, target) {
    // if(reader.get(0) == Integer.MAX_VALUE) return -1;
    let low = 0, high = 1;
    while(reader.get(high) < target){
        low = high;
        high = high * 2;
    }
    return find_element(reader,target,low,high);
};

function find_element(reader,target,low,high){
    while(low<=high){
        let mid = low + Math.floor((high-low)/2);
        if(reader.get(mid) == target) {
            return mid;
        }
        else if(target > reader.get(mid)){
            low = mid + 1;
        }else{
            high = mid -1;
        }
    }
    return -1;
}